# Complete Python Bootcamp: Go from Hero to Zero
updated: Friday 13th 1:15

## Part 1:

\begin{table}[]
\begin{tabular}{lll}
Name           & Type  & Description                                                      \\
Integers       & int   & Whole numbers, such as: 3 300 200                                \\
Floating point & float & Numbers with a decimal point: 2.3 4.6 100.0                      \\
Strings        & str   & Ordered sequence of characters: "hello" 'Sammy' "2000" "chinese" \\
Lists          & list  & Ordered sequences of objects: {[}10,"hello",200.3{]}             \\
Dictionaries   & dict  & Unordered Key:Value pairs: \{"mykey":"value","name":"Frankie"\}  \\
Tuples         & tup   & Ordered immutable sequence of objects: (10,"hello",200.3)        \\
Sets           & set   & Unordered collection of unique objects: \{"a","b"\}              \\
Booleans       & bool  & Logical value indicating True or False                          
\end{tabular}
\end{table}

## 1.Numbers

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

### Types of numbers

- Many "types" of numbers (numeric literals). 
- Focus on integers (1) and floating point numbers (1.0).

- **Integers** are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

- **Floating point numbers** have a decimal point in them, or use an exponential (e) to define the number. 
    - 2.0 and -2.1 are examples of floating point numbers. 
    - 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

##### Arithmetic

In [6]:
# Add
1+2

3

In [7]:
# Subtract
1-2

-1

In [8]:
# Multiplication
1*2

2

In [9]:
# Division
10/2

5.0

In [13]:
# Floor Division
7//4

1

###### Modulo / "Mod" operator

In [17]:
7/4 # Division

1.75

In [18]:
# Floor Division
7//4

1

**7 divided by 4 equals 1.75 not 1!**

Use modulus to check if a number is even or odd. The remainder value.

In [19]:
7%4

3

In [20]:
10%2 # anything mod 2 is equal to 0 if even 

0

In [22]:
11%2 # anything mod 2 is equal to 1 if odd

1

In [23]:
# Exponents /Powers
2**3  

8

In [24]:
# Can also do roots this way
4**0.5

2.0

In [25]:
# BODMAS
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [26]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

#### Variable Assignments

Names you use when creating these labels need to follow a few rules:

    1. Names can not start with a number.
    2. There can be no spaces in the name, use _ instead.
    3. Can't use any of these symbols :'",<>/?|\()!@#$%^&*~-+
    4. It's considered best practice (PEP8) that names are lowercase.
    5. Avoid using the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), 
       or 'I' (uppercase letter eye) as single character variable names.
    6. Avoid using words that have special meaning in Python like "list" and "str"


Using variable names can be a very useful way to keep track of different variables in Python.

#### Rules for variable names
* names can not start with a number
* names can not contain spaces, use _ intead
* names can not contain any of these symbols:

      :'",<>/?|\!@#%^&*~-+
       
* it's considered best practice ([PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are lowercase with underscores
* avoid using Python built-in keywords like `list` and `str`
* avoid using the single characters `l` (lowercase letter el), `O` (uppercase letter oh) and `I` (uppercase letter eye) as they can be confused with `1` and `0`

#### Dynamic Typing

- Python uses **dynamic typing**, meaning you can reassign variables to different data types. 
- This makes Python very flexible in assigning data types; it differs from other languages that are *statically typed*.

##### Pros of Dynamic Typing
* very easy to work with
* faster development time

##### Cons of Dynamic Typing
* may result in unexpected bugs!
* you need to be aware of `type()`

##### Determining variable type with `type()`
You can check what type of object is assigned to a variable using Python's built-in `type()` function. Common data types include:
* **int** (for integer)
* **float**
* **str** (for string)
* **list**
* **tuple**
* **dict** (for dictionary)
* **set**
* **bool** (for Boolean True/False)

## 2.  Strings


    1.) Creating Strings
    2.) Printing Strings
    3.) String Indexing and Slicing
    4.) String Properties
    5.) String Methods
    6.) Print Formatting

In [35]:
# Single word
'hello'

'hello'

In [36]:
# We can also use double quote
"String built with double quotes"

'String built with double quotes'

In [37]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-37-da9a34b3dc31>, line 2)

The reason for the error above is because the single quote in <code>I'm</code> stopped the string. You can use combinations of double and single quotes to get the complete statement.

#### Printing a String

Using Jupyter notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a print function.

In [39]:
#### Declare a string
'Hello World'

'Hello World'

In [40]:
# Note that we can't output multiple strings this way
'Hello World 1'
'Hello World 2'

'Hello World 2'

In [42]:
## Print Multiple Strings
print('Hello World 1')
print('Hello World 2')

print('Use \n to print a new line')
print('\n')
print('Cool Thamu')

Hello World 1
Hello World 2
Use 
 to print a new line


Cool Thamu


##### String Basics
- use a function called len() to check the length of a string!
- Python's built-in len() function counts all of the characters in the string, including spaces and punctuation.

In [45]:
len('Man')

3

In [48]:
'Man'[0]

'M'

In [50]:
'Man'[1]

'a'

### String Indexing
We know strings are a sequence, which means Python can use indexes to call parts of the sequence. Let's learn how this works.

In Python, we use brackets <code>[]</code> after an object to call its index. We should also note that indexing starts at 0 for Python. Let's create a new object called <code>s</code> and then walk through a few examples of indexing.

- Slicing allows you to grab a subsection of multiple characters, a "slice" of the string.
- **[start:stop:step]**
- **start** is a numerical index for the slice start
- **stop** is the index you will go up to (but not include)
- **step** is the size of the "jump" you take.

In [59]:
print("Zimbabwe is great"[0])
print("Zimbabwe is great"[1])
print("Zimbabwe is great"[5:])
print("\n \n Zimbabwe is \t great"[5:])

Z
i
bwe is great
imbabwe is 	 great


In [61]:
print("Zimbabwe is great"[:3])

Zim


In [65]:
print("Zimbabwe is great"[0:8])

Zimbabwe


We can use a <code>:</code> to perform *slicing* which grabs everything up to a designated point. For example:

In [66]:
print("Zimbabwe is great"[::])

Zimbabwe is great


In [67]:
print("Zimbabwe is great"[::2])

Zmaw sget


In [68]:
print("abcdefg"[::2])

aceg


In [70]:
print("abcdefg"[::-1]) # Reverse

gfedcba


In [71]:
# Assign s as a string
s = 'Hello World'

In [72]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

- Note the above slicing. Here we're telling Python to grab everything from 0 up to 3. 
- It doesn't include the 3rd index. 
- You'll notice this a lot in Python, where **statements and are usually in the context of "up to, but not including"**.

- We can also use negative indexing to go backwards.

In [74]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [75]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [76]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [77]:
# We can use this to print a string backwards
s[::-1]

'dlroW olleH'

In [78]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

#### String Properties
Important to note that **strings have an important property** known as **immutability**. 
- means that **once a string is created, the elements within it can not be changed or replaced**. 

**Immutability**:

In [79]:
s

'Hello World'

In [80]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Notice how the error tells us directly what we can't do, change the item assignment!

Something we **can do is concatenate strings**!

In [84]:
s

'Hello World'

In [86]:
 s[6:11]

'World'

In [89]:
last_name =  s[6:11]
new_name = "Fred's " + last_name
new_name

"Fred's World"

In [90]:
"A" * 10

'AAAAAAAAAA'

In [91]:
# Upper Case a string
s.upper()

'HELLO WORLD'

In [92]:
# Lower case
s.lower()

'hello world'

In [93]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'World']

In [98]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld']

In [99]:
s.split(' ')

['Hello', 'World']

In [104]:
# Split by a specific element (doesn't include the element that was split on)
new_name.split('r')

['F', "ed's Wo", 'ld']

There are many more methods than the ones covered here. Visit the Advanced String section to find out more!

---
1) Are strings Mutable?

- Stings are not mutable. Meaning, you can't use indexing to change individual elements of a string.

2) How do I create comments in my code?

- You can use the hashtag # to create comments in your code.
---

#### Print Formatting

We can use the .format() method to add formatted objects to printed string statements. 

- **String interpolation** - often you will want to "inject" a variable into your string for printing.
- Multiple ways to format strings for printing variables in them.
The easiest way to show this is through an example:

In [97]:
s.format()

'Hello World'

In [105]:
'Sammy'[2:]

'mmy'

In [106]:
'Insert another string with curly brackets: {}'.format('The inserted string')

'Insert another string with curly brackets: The inserted string'

### Next up: Lists!

## 3. Print formatting with Strings

Let's explore two methods for string interpolation:

1) **.format()** method

2) **f-strings** (formatted string literals)

#### Formatting with the .format() method

To format objects into your stings for print statements is with the string .format()
method. The syntax is:
    
 **'String here {} then also {}'.format('something1','something2')**

In [121]:
print('This is a {}'.format("Monday","Tuesday","Wednesday"))

This is a Monday


In [123]:
print('This is a {0}'.format("Monday","Tuesday","Wednesday"))
print('This is a {1}'.format("Monday","Tuesday","Wednesday"))
print('This is a {2}'.format("Monday","Tuesday","Wednesday"))

This is a Monday
This is a Tuesday
This is a Wednesday


In [126]:
print('I love {1} {2} {0}'.format("Monday","Tuesday","Wednesday"))

I love Tuesday Wednesday Monday


In [127]:
print('I love {0}{0}{0}'.format("Monday","Tuesday","Wednesday"))

I love MondayMondayMonday


In [129]:
# Use Keywords
print('I love {M} {W} {T}'.format(M="Monday",T="Tuesday",W="Wednesday"))

I love Monday Wednesday Tuesday


----
**Float formatting follows "{value:width.precision f}"**

In [131]:
result = 100/777
result 

0.1287001287001287

In [132]:
print("The result is equal to {}".format(result))

The result is equal to 0.1287001287001287


In [142]:
print("The result is equal to {r}".format(r=result))

The result is equal to 0.1287001287001287


In [145]:
# Change the width 
print("The result is equal to {r:1.3f}".format(r=result))
print("The result is equal to {r:10.3f}".format(r=result))

The result is equal to 0.129
The result is equal to      0.129


In [143]:
# Change the precision
print("The result is equal to {r:1.3f}".format(r=result))
print("The result is equal to {r:1.10f}".format(r=result))

The result is equal to 0.129
The result is equal to 0.1287001287


String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them here.

In [128]:
# Look at our example
player = 'Thomas'
points = 33
'Last night, '+player+' scored '+str(points)+' points.'  # concatenation

'Last night, Thomas scored 33 points.'

In [117]:
f'Last night, {player} scored {points} points.'          # string formatting

'Last night, Thomas scored 33 points.'

In [119]:
'Last night, {} scored {} points.'.format(player,points)   

'Last night, Thomas scored 33 points.'

#### Formatting with placeholders
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [120]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


##### Padding and Precision of Floating Point Numbers " %width.precision f "
- Floating point numbers use the format <code>%5.2f</code>. 
- Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. 
- Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point.

In [146]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [154]:
# Change the precision
print('Floating point numbers: %1.0f' %(13.144))
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13
Floating point numbers: 13.14400


In [153]:
# Change the width
print('Floating point numbers: %5.2f' %(13.144))
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers: 13.14
Floating point numbers:      13.14


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Multiple Formatting
Nothing prohibits using more than one conversion tool in the same print statement:

In [157]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


#### Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
   

In [159]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


#### The .format() method has several advantages over the %s placeholder method:

##### 1. Inserted objects can be called by index position:

In [162]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


##### 2. Inserted objects can be assigned keywords:

In [164]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


##### 3. Inserted objects can be reused, avoiding duplication:

In [166]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


##### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [168]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [169]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [171]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [173]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

#### Formatted String Literals (f-strings)